File: src/main/java/com/mycompany/model/Product.java
package com.mycompany.model;

import com.mycompany.util.Constants;
import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Class Product - Aggregation dengan Category
 */
public class Product {
    private int id;
    private String sku;
    private String name;
    private Category category; // Aggregation
    private BigDecimal price;
    private int stock;
    private String description;
    private int discountPercentage; // Diskon produk dalam persen (0-100)
    
    // Constructor
    public Product() {}
    
    public Product(int id, String sku, String name, Category category, 
                   BigDecimal price, int stock, String description) {
        this.id = id;
        this.sku = sku;
        this.name = name;
        this.category = category;
        this.price = price;
        this.stock = stock;
        this.description = description;
    }
    
    // Encapsulation - Getters and Setters
    public int getId() { return id; }
    public void setId(int id) { this.id = id; }
    
    public String getSku() { return sku; }
    public void setSku(String sku) { this.sku = sku; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public Category getCategory() { return category; }
    public void setCategory(Category category) { this.category = category; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) {
        if (price.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Price cannot be negative");
        }
        this.price = price;
    }
    
    public int getStock() { return stock; }
    public void setStock(int stock) {
        if (stock < 0) {
            throw new IllegalArgumentException("Stock cannot be negative");
        }
        this.stock = stock;
    }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public int getDiscountPercentage() { return discountPercentage; }
    public void setDiscountPercentage(int discountPercentage) {
        if (discountPercentage < Constants.MIN_DISCOUNT_PERCENTAGE || discountPercentage > Constants.MAX_DISCOUNT_PERCENTAGE) {
            throw new IllegalArgumentException("Discount must be between 0 and 100");
        }
        this.discountPercentage = discountPercentage;
    }
    
    // Business method
    public BigDecimal getSubtotal(int quantity) {
        // Hitung subtotal tanpa diskon per-produk
        return price.multiply(new BigDecimal(quantity));
    }
    
    // Hitung harga setelah diskon
    public BigDecimal getPriceAfterDiscount() {
        // Hitung harga satuan setelah diskon per-produk
        if (discountPercentage > 0) {
            BigDecimal discount = price.multiply(new BigDecimal(discountPercentage))
                                      .divide(new BigDecimal(100), 2, RoundingMode.HALF_UP);
            return price.subtract(discount);
        }
        return price;
    }
    
    // Hitung subtotal dengan diskon
    public BigDecimal getSubtotalWithDiscount(int quantity) {
        // Hitung subtotal setelah diskon per-produk
        return getPriceAfterDiscount().multiply(new BigDecimal(quantity));
    }
    
    public boolean isAvailable(int quantity) {
        // Cek stok cukup untuk quantity tertentu
        return stock >= quantity;
    }
    
    public void reduceStock(int quantity) {
        // Kurangi stok (validasi stok cukup)
        if (!isAvailable(quantity)) {
            throw new IllegalStateException("Insufficient stock");
        }
        this.stock -= quantity;
    }
    
    public void addStock(int quantity) {
        // Tambah stok produk
        this.stock += quantity;
    }
}

File: ciomart_flutter/lib/models/product.dart
import 'category.dart';

class Product {
  final int? id;
  final String sku;
  final String name;
  final double price;
  int stock;
  final int categoryId;
  final Category? category;
  final int discountPercentage;

  Product({
    this.id,
    required this.sku,
    required this.name,
    required this.price,
    required this.stock,
    required this.categoryId,
    this.category,
    this.discountPercentage = 0,
  });

  // Business Logic Methods
  double getPriceAfterDiscount() {
    if (discountPercentage > 0) {
      return price * (100 - discountPercentage) / 100;
    }
    return price;
  }

  bool isAvailable(int quantity) {
    return stock >= quantity;
  }

  void reduceStock(int quantity) {
    if (isAvailable(quantity)) {
      stock -= quantity;
    } else {
      throw Exception('Stok tidak cukup');
    }
  }

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id']?.toString() ?? ''),
      sku: json['sku'],
      name: json['name'],
      price: json['price'] is num ? (json['price'] as num).toDouble() : double.parse(json['price'].toString()),
      stock: json['stock'] is int ? json['stock'] : int.parse(json['stock'].toString()),
      categoryId: json['category_id'] is int ? json['category_id'] : int.parse(json['category_id'].toString()),
      category: json['category_name'] != null 
          ? Category(
              id: json['category_id'] is int ? json['category_id'] : int.parse(json['category_id'].toString()), 
              name: json['category_name']) 
          : null,
      discountPercentage: json['discount_percentage'] is int 
          ? json['discount_percentage'] 
          : int.tryParse(json['discount_percentage']?.toString() ?? '0') ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'sku': sku,
      'name': name,
      'price': price,
      'stock': stock,
      'category_id': categoryId,
      'discount_percentage': discountPercentage,
    };
  }
}
